### ISSUE TITLE
Rule L028 fails when using keywords in a sql-function in a query containing joins

### ISSUE THREAD

**Issue by johansjob** [sqlfluff/sqlfluff#357]
**Problem**
When using keywords like 'day', 'month' etc. in a sql-function in a query containing joins.
It treats the keywords like a column name and expects a qualified name. 

**Example**
```
with a as (
    select 1 as id, cast('2020-01-01' as timestamp) as ts
)

, b as (
    select 2 as id, 'some_field' as some_field
)

select
    timestamp_trunc(a.ts, month) as t
from a
full join b using (id)
```

**Expected**
Keywords should be recognised and not result in an unqualified error.

**Settings**
Version: 0.3.4
Dialect: BigQuery

**Comment by alanmcruickshank** [comment]
Good find, thank you for raising this one. This could be a tricky one.

I don't really want to start hard coding function signatures into the parser, but I'm not sure how we can always conclusively know whether to expect a date-part or whether to expect a reference.

Likely any approaches which *don't* involve hard coding specific function signatures will involve some false alarms on either side. Either occasionally we mistake an identifier for a date-part or visa versa. On balance, given we've got less linting around date-parts, occasionally mistaking identifiers for date-parts feels better than the other way around.

**Comment by pwildenhain** [comment]
Should this be added to the changes made in #424 or #443? Or is this separate from that?

Linting this query no longer triggers `L028` but does violate `L027`
test.sql
```sql
with a as (
    select 1 as id, cast('2020-01-01' as timestamp) as ts
),

b as (
    select 2 as id, 'some_field' as some_field
)

select
    timestamp_trunc(a.ts, month) as t
from a
full join b using (id)
```

```sh
$ sqlfluff lint test.sql --dialect bigquery
== [test.sql] FAIL
L:  10 | P:  27 | L027 | Unqualified reference 'month' found in select with more
                       | than one referenced table/view.
```

**Comment by barrywhart** [comment]
This seems _slightly_ related to #315, where date part keywords are being flagged as bad column aliases.

I wonder... would it be reasonable for L027, before producing a lint warning, to see if the word is a keyword, and if so, do not emit a warning?

Pros:
* Avoids hardcoding function signatures into SQLFluff
* Could, if needed, handle keywords to vary per dialect-specific

Cons:
* Will allow use of keywords in places where it shouldn't.

To help reduce the con above:
* Maybe introduce a separate, narrower list of "keywords that can appear in function calls" or something like that

**Comment by barrywhart** [comment]
Is this potentially related to #430? The keywords we're discussing here behave a _bit_ like bare functions.

**Comment by barrywhart** [comment]
Resolved by #708
